// twilio/dist/lib/rest/flexApi/V1.js
"use strict";
/*
 * This code was generated by
 * ___ _ _ _ _ _    _ ____    ____ ____ _    ____ ____ _  _ ____ ____ ____ ___ __   __
 *  |  | | | | |    | |  | __ |  | |__| | __ | __ |___ |\ | |___ |__/ |__|  | |  | |__/
 *  |  |_|_| | |___ | |__|    |__| |  | |    |__] |___ | \| |___ |  \ |  |  | |__| |  \
 *
 * Twilio - Flex
 * This is the public Twilio REST API.
 *
 * NOTE: This class is auto generated by OpenAPI Generator.
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Version_1 = __importDefault(require("../../base/Version"));
const assessments_1 = require("./v1/assessments");
const channel_1 = require("./v1/channel");
const configuration_1 = require("./v1/configuration");
const flexFlow_1 = require("./v1/flexFlow");
const insightsQuestionnairesCategory_1 = require("./v1/insightsQuestionnairesCategory");
const insightsQuestionnairesQuestion_1 = require("./v1/insightsQuestionnairesQuestion");
const insightsSession_1 = require("./v1/insightsSession");
const insightsSettingsAnswerSets_1 = require("./v1/insightsSettingsAnswerSets");
const insightsSettingsComment_1 = require("./v1/insightsSettingsComment");
const insightsUserRoles_1 = require("./v1/insightsUserRoles");
const interaction_1 = require("./v1/interaction");
const webChannel_1 = require("./v1/webChannel");
class V1 extends Version_1.default {
    /**
     * Initialize the V1 version of FlexApi
     *
     * @param domain - The Twilio (Twilio.FlexApi) domain
     */
    constructor(domain) {
        super(domain, "v1");
    }
    /** Getter for assessments resource */
    get assessments() {
        this._assessments = this._assessments || (0, assessments_1.AssessmentsListInstance)(this);
        return this._assessments;
    }
    /** Getter for channel resource */
    get channel() {
        this._channel = this._channel || (0, channel_1.ChannelListInstance)(this);
        return this._channel;
    }
    /** Getter for configuration resource */
    get configuration() {
        this._configuration =
            this._configuration || (0, configuration_1.ConfigurationListInstance)(this);
        return this._configuration;
    }
    /** Getter for flexFlow resource */
    get flexFlow() {
        this._flexFlow = this._flexFlow || (0, flexFlow_1.FlexFlowListInstance)(this);
        return this._flexFlow;
    }
    /** Getter for insightsQuestionnairesCategory resource */
    get insightsQuestionnairesCategory() {
        this._insightsQuestionnairesCategory =
            this._insightsQuestionnairesCategory ||
                (0, insightsQuestionnairesCategory_1.InsightsQuestionnairesCategoryListInstance)(this);
        return this._insightsQuestionnairesCategory;
    }
    /** Getter for insightsQuestionnairesQuestion resource */
    get insightsQuestionnairesQuestion() {
        this._insightsQuestionnairesQuestion =
            this._insightsQuestionnairesQuestion ||
                (0, insightsQuestionnairesQuestion_1.InsightsQuestionnairesQuestionListInstance)(this);
        return this._insightsQuestionnairesQuestion;
    }
    /** Getter for insightsSession resource */
    get insightsSession() {
        this._insightsSession =
            this._insightsSession || (0, insightsSession_1.InsightsSessionListInstance)(this);
        return this._insightsSession;
    }
    /** Getter for insightsSettingsAnswerSets resource */
    get insightsSettingsAnswerSets() {
        this._insightsSettingsAnswerSets =
            this._insightsSettingsAnswerSets ||
                (0, insightsSettingsAnswerSets_1.InsightsSettingsAnswerSetsListInstance)(this);
        return this._insightsSettingsAnswerSets;
    }
    /** Getter for insightsSettingsComment resource */
    get insightsSettingsComment() {
        this._insightsSettingsComment =
            this._insightsSettingsComment ||
                (0, insightsSettingsComment_1.InsightsSettingsCommentListInstance)(this);
        return this._insightsSettingsComment;
    }
    /** Getter for insightsUserRoles resource */
    get insightsUserRoles() {
        this._insightsUserRoles =
            this._insightsUserRoles || (0, insightsUserRoles_1.InsightsUserRolesListInstance)(this);
        return this._insightsUserRoles;
    }
    /** Getter for interaction resource */
    get interaction() {
        this._interaction = this._interaction || (0, interaction_1.InteractionListInstance)(this);
        return this._interaction;
    }
    /** Getter for webChannel resource */
    get webChannel() {
        this._webChannel = this._webChannel || (0, webChannel_1.WebChannelListInstance)(this);
        return this._webChannel;
    }
}
exports.default = V1;
